ssa/manager_apply.go

```diff
@@ -127,58 +128,85 @@ func (m *ResourceManager) Apply(ctx context.Context, object *unstructured.Unstru
 	sort.Sort(SortableUnstructureds(objects))
-	changeSet := NewChangeSet()
-	var toApply []*unstructured.Unstructured
-	for _, object := range objects {
-		existingObject := object.DeepCopy()
-		getError := m.client.Get(ctx, client.ObjectKeyFromObject(object), existingObject)
-
-		if m.shouldSkipApply(object, existingObject, opts) {
-			changeSet.Add(*m.changeSetEntry(existingObject, SkippedAction))
-			continue
-		}
 
-		dryRunObject := object.DeepCopy()
-		if err := m.dryRunApply(ctx, dryRunObject); err != nil {
-			// we cannot have an immutable error (and therefore shouldn't false apply) if the resource doesn't exist
-			// on the cluster. Note that resource might not exist because we wrongly identified an error as immutable and deleted
-			// it when ApplyAll was called the last time (the check for ImmutableError returns false positives)
-			if !errors.IsNotFound(getError) && m.shouldForceApply(object, existingObject, opts, err) {
-				if err := m.client.Delete(ctx, existingObject); err != nil && !errors.IsNotFound(err) {
-					return nil, fmt.Errorf("%s immutable field detected, failed to delete object: %w",
-						FmtUnstructured(dryRunObject), err)
+	// Results are written to the following arrays from the concurrent goroutines. We use arrays
+	// to avoid complex synchronization. toApply is sparse, slots are only popuplated when there
+	// is an object to apply
+	toApply := make([]*unstructured.Unstructured, len(objects))
+	changes := make([]ChangeSetEntry, len(objects))
+
+	{
+		g, ctx := errgroup.WithContext(ctx)
+		g.SetLimit(m.concurrency)
+		for i, object := range objects {
+			i, object := i, object
+
+			g.Go(func() error {
+				existingObject := object.DeepCopy()
+				getError := m.client.Get(ctx, client.ObjectKeyFromObject(object), existingObject)
+
+				if m.shouldSkipApply(object, existingObject, opts) {
+					changes[i] = *m.changeSetEntry(existingObject, SkippedAction)
+					return nil
 				}
-				return m.ApplyAll(ctx, objects, opts)
-			}
 
-			return nil, m.validationError(dryRunObject, err)
-		}
+				dryRunObject := object.DeepCopy()
+				if err := m.dryRunApply(ctx, dryRunObject); err != nil {
+					// we cannot have an immutable error (and therefore shouldn't false apply) if the resource doesn't exist
+					// on the cluster. Note that resource might not exist because we wrongly identified an error as immutable and deleted
+					// it when ApplyAll was called the last time (the check for ImmutableError returns false positives)
+					if !errors.IsNotFound(getError) && m.shouldForceApply(object, existingObject, opts, err) {
+						if err := m.client.Delete(ctx, existingObject); err != nil && !errors.IsNotFound(err) {
+							return fmt.Errorf("%s immutable field detected, failed to delete object: %w",
+								FmtUnstructured(dryRunObject), err)
+						}
+
+						// retry dry run
+						err = m.dryRunApply(ctx, dryRunObject)
```

Comment:
Retrying only once will always fail for objects with finalizers because `client.Delete` is non-blocking. For example, we delete a deployment with changes to the pod label selectors, the deployment will be stuck for a long time in finalization until all pods and volumes are removed.

Reply:
Thanks, I see. In that case, a retry loop with intentional delays, instead of unbounded recursion only delayed by QPS and CPU might be better? Something like this?

```diff
index 5d7e524..5f0526b 100644
--- a/ssa/manager_apply.go
+++ b/ssa/manager_apply.go
@@ -28,6 +28,7 @@ import (
        "k8s.io/apimachinery/pkg/api/errors"
        "k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
        "k8s.io/apimachinery/pkg/types"
+       "k8s.io/apimachinery/pkg/util/wait"
        "sigs.k8s.io/controller-runtime/pkg/client"
 )
 
@@ -161,7 +162,31 @@ func (m *ResourceManager) ApplyAll(ctx context.Context, objects []*unstructured.
                                                                FmtUnstructured(dryRunObject), err)
                                                }
 
-                                               // retry dry run
+                                               // 1s, 2s, 4, 8, 16, 32, 1m, 1m, 1m, 1m, 1m, 1m, 1m, 1m, 1m
+                                               backoff := wait.Backoff{
+                                                       Duration: 1 * time.Second,
+                                                       Factor:   2,
+                                                       Steps:    15,
+                                                       Cap:      1 * time.Minute,
+                                               }
+
+                                               // wait until deleted (in case of any finalizers)
+                                               err = wait.ExponentialBackoffWithContext(ctx, backoff, func(ctx context.Context) (bool, error) {
+                                                       err := m.client.Get(ctx, client.ObjectKeyFromObject(object), existingObject)
+                                                       switch {
+                                                       case err == nil:
+                                                               return false, nil // object still exists
+                                                       case errors.IsNotFound(err):
+                                                               return true, nil // done
+                                                       default:
+                                                               return false, err // other error
+                                                       }
+                                               })
+                                               if err != nil {
+                                                       return fmt.Errorf("%s immutable field detected, failed to wait for object to be deleted: %w",
+                                                               FmtUnstructured(dryRunObject), err)
+                                               }
+
                                                err = m.dryRunApply(ctx, dryRunObject)
                                        }
```

It would also be nice with an integration test for this, I'll see if I can come up with one.

Reply:
Ah that’s no good, we need a variant of `WaitForTermination` but with a context arg to the whole thing timeouts according to the apply deadline.

Reply:
I added an integration test for this case:

```go
	t.Run("force apply waits for finalizer", func(t *testing.T) {
		secretClone := secret.DeepCopy()
		{
			secretWithFinalizer := secretClone.DeepCopy()

			unstructured.SetNestedStringSlice(secretWithFinalizer.Object, []string{"fluxcd.io/demo-finalizer"}, "metadata", "finalizers")
			if err := manager.client.Update(ctx, secretWithFinalizer); err != nil {
				t.Fatal(err)
			}
		}

		// remove finalizer after a delay, to ensure the controller handles a slow deletion
		go func() {
			time.Sleep(3 * time.Second)

			secretWithoutFinalizer := secretClone.DeepCopy()
			unstructured.SetNestedStringSlice(secretWithoutFinalizer.Object, []string{}, "metadata", "finalizers")
			if err := manager.client.Update(ctx, secretWithoutFinalizer); err != nil {
				panic(err)
			}
		}()

		// update a value in the secret
		err = unstructured.SetNestedField(secret.Object, "val-secret2", "stringData", "key")
		if err != nil {
			t.Fatal(err)
		}

		// force apply
		changeSet, err := manager.ApplyAllStaged(ctx, objects, ApplyOptions{Force: true, WaitTimeout: timeout})
		if err != nil {
			t.Fatal(err)
		}

		// verify the secret was recreated
		for _, entry := range changeSet.Entries {
			if entry.Subject == secretName {
				if diff := cmp.Diff(CreatedAction, entry.Action); diff != "" {
					t.Errorf("Mismatch from expected value (-want +got):\n%s", diff)
				}
			} else {
				if diff := cmp.Diff(UnchangedAction, entry.Action); diff != "" {
					t.Errorf("Mismatch from expected value (-want +got):\n%s", diff)
				}
			}
		}
	})
```

Reply:
Instead of setting up the backoff on our own we can use `wait.PollUntilContextCancel`.